src/cdomains/lval.ml

```diff
@@ -324,10 +386,32 @@ end
     - {!NullPtr} is a singleton sublattice.
     - {!UnknownPtr} is a singleton sublattice.
     - If [ana.base.limit-string-addresses] is enabled, then all {!StrPtr} are together in one sublattice with flat ordering. If [ana.base.limit-string-addresses] is disabled, then each {!StrPtr} is a singleton sublattice. *)
-module NormalLat (Idx: IdxDomain) =
+module NormalLat (Idx: IntDomain.Z) =
 struct
   include Normal (Idx)
-  module Offs = Offset (Idx)
+  module Offs = OffsetLatWithSemanticEqual (Idx)
+
+  let semantic_equal x y = match x, y with
+    | Addr (x, xoffs), Addr (y, yoffs) ->
+      if CilType.Varinfo.equal x y then
+        let xtyp = x.vtype in
+        let ytyp = y.vtype in
+        Offs.semantic_equal ~xtyp ~xoffs ~ytyp ~yoffs
+      else
+        Some false
+    | StrPtr None, StrPtr _
+    | StrPtr _, StrPtr None -> Some true
+    | StrPtr (Some a), StrPtr (Some b) -> Some (a = b)
+    | NullPtr, NullPtr -> Some true
+    | UnknownPtr, UnknownPtr -> None
+    | UnknownPtr, Addr _
+    | Addr _, UnknownPtr
+    | UnknownPtr, StrPtr _
+    | StrPtr _, UnknownPtr ->
+      (* TODO: Case needed? *)
+      None
+    | _, _ ->
+      Some false
```

Comment:
What does the `bool option` mean for semantic equality here?

Reply:
`Some true` means they are definitely equal, `Some false`  means they are definitely not equal, and `None` means they may be equal. Here, unknown pointers may be equal to everything.